softwareComponents/torqueComputation/src/joint.cpp

```diff
@@ -0,0 +1,212 @@
+             canonicNeighborsIds.begin(),
+             canonicNeighborsIds.end(), 
+            neighbor.id) ==  canonicNeighborsIds.end()
+        ) {
+            throw std::logic_error("Provided joint is not a neighbor of handled joint");
+        }
+
+        vec  neighborVector = neighbor.coors - coors;
+        return projectMatrixOntoVector(moments,  neighborVector);
+    }
+
+    void Joint::appendNeighbor(Joint* neighbor) {
+        neighbors.push_back(neighbor);
+    }
+
+    void Joint::appendForce(arma::vec3 force) {
+        forces.push_back(force);   
+    }
+
+    Joint* Joint::getNeighbourById(int  neighborId) {
+        for (Joint* neighbor : neighbors) {
+            if (neighbor->id ==  neighborId) {
+                return neighbor;
+            }
+        }
+        return nullptr;
+    }
+
+    bool Joint::jointSanityCheck(std::string& msg) const {
+        if (!isBounded) {
+            return true;
+        }
+
+        if (2 < neighbors.size()) {
+            msg = "Bounded motor can only have at most 2 neighbors";
+            return false;
+        }
+
+        for (Joint* neighbor : neighbors) {
+            vec linkVector = neighbor->coors - coors;
+            //Not perpendicular nor parallel is not valid for main moment
+            if (!almostEqual(dot(moments.row(2), linkVector), 0.0) &&
+                !isParallel(moments.row(2).t(), linkVector)) {
+                    std::stringstream stream;
+                    stream << "Joint is in invalid state!" << std::endl <<
+                              "ID: " << id << std::endl <<
+                              "Perpendicular: " << dot(moments.row(2), linkVector) << " == 0.0" << std::endl <<
+                              "Norm: " << moments.row(2) <<
+                              "Link vector: " << linkVector.t() << std::endl;
+
+                    msg = stream.str();
+                    return false;
+            }
+        }
+        return true;
+    }
+
+    std::vector<int> Joint::getCanonicNeighborsIds() const {
+        std::vector<int>  neighborIds;
+        for (Joint* neighbor : neighbors) {
+             neighborIds.push_back(neighbor->id);
+            
+        }
+        std::sort( neighborIds.begin(),  neighborIds.end());
+        return  neighborIds;
+    }
+
+    bool Joint::operator==(const Joint& other) const {
+        const double TOLERANCE = 1e-6;
+        return id == other.id &&
+            approx_equal(coors, other.coors, "absdiff", TOLERANCE) &&
+            (
+                (norm == std::nullopt && other.norm == std::nullopt) ||
+                (
+                    norm && other.norm &&
+                    approx_equal(norm.value(), other.norm.value(), "absdiff", TOLERANCE)
+                )
+            ) &&
+            isWall == other.isWall &&
+            isBounded == other.isBounded &&
+            approx_equal(moments, other.moments, "absdiff", TOLERANCE);
+    }
+
+    size_t Joint::operator()() const {
```

Comment:
Jaká je motivace za tím, aby `operator()` počítal hash? Proč se to prostě nejmenuje `hash`?